unified-runtime/source/adapters/level_zero/v2/queue_immediate_in_order.cpp

```diff
@@ -712,25 +716,106 @@ ur_result_t ur_queue_immediate_in_order_t::enqueueWriteHostPipe(
+    commandType = UR_COMMAND_ENQUEUE_USM_DEVICE_ALLOC_EXP;
+    break;
+  case UR_USM_TYPE_SHARED:
+    commandType = UR_COMMAND_ENQUEUE_USM_SHARED_ALLOC_EXP;
+    break;
+  default:
+    logger::error("enqueueUSMAllocHelper: unsupported USM type");
+    throw UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  auto zeSignalEvent = getSignalEvent(commandListLocked, phEvent, commandType);
+  auto [pWaitEvents, numWaitEvents] =
+      getWaitListView(commandListLocked, phEventWaitList, numEventsInWaitList);
+
+  if (numWaitEvents > 0) {
+    ZE2UR_CALL(
+        zeCommandListAppendWaitOnEvents,
+        (commandListLocked->getZeCommandList(), numWaitEvents, pWaitEvents));
+  }
+
+  if (zeSignalEvent) {
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (commandListLocked->getZeCommandList(), zeSignalEvent));
+  }
+
+  return UR_RESULT_SUCCESS;
+}
+
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMDeviceAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_queue_handle_t hQueue, ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  return enqueueUSMAllocHelper(hQueue, pPool, size, pProperties,
+                               numEventsInWaitList, phEventWaitList, ppMem,
+                               phEvent, UR_USM_TYPE_DEVICE);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_queue_handle_t hQueue, ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  return enqueueUSMAllocHelper(hQueue, pPool, size, pProperties,
+                               numEventsInWaitList, phEventWaitList, ppMem,
+                               phEvent, UR_USM_TYPE_SHARED);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_queue_handle_t hQueue, ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  return enqueueUSMAllocHelper(hQueue, pPool, size, pProperties,
```

Comment:
Please add `TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp");`  at the beginning of this function (and other function you implement in this patch).